src/tests/suite/daos_nvme_recovery.c

```diff
@@ -94,28 +125,101 @@ nvme_recov_1(void **state)
 	}
 	ioreq_fini(&req);
 
-	print_message("Error injection to simulate device faulty.\n");
-	set_fail_loc(arg, rank, DAOS_NVME_FAULTY | DAOS_FAIL_ONCE);
+	/** Query test args to get total pool target count per node */
+	assert_true(arg->srv_ntgts > arg->srv_nnodes);
+	per_node_tgt_cnt = arg->srv_ntgts/arg->srv_nnodes;
 
-	/*
-	 * FIXME: Due to lack of infrastructures for checking each target
-	 *	  status, let's just wait for an arbitrary time and hope the
-	 *	  faulty reaction & rebuild is triggered.
+	/**
+	 * Verify initial states for all pool targets are UPIN by querying
+	 * the pool target info.
 	 */
+	for (i = 0; i < per_node_tgt_cnt; i++) {
+		rc = daos_pool_query_target(arg->pool.poh, i/*tgt*/,
+					    rank, &tgt_info, NULL /*ev*/);
+		assert_int_equal(rc, 0);
+		rc = strcmp(daos_target_state_enum_to_str(tgt_info.ta_state),
+			    "UPIN");
+		assert_int_equal(rc, 0);
+	}
+	print_message("All targets are in UPIN\n");
+
+	/** Inject error on random target index */
+	srand(time(NULL));
+	fail_loc_tgt = rand() % per_node_tgt_cnt;
+	print_message("Error injection on tgt %"PRIu64" to simulate device"
+		      " faulty.\n", fail_loc_tgt);
+	set_fail_loc(arg, rank, fail_loc_tgt, DAOS_NVME_FAULTY | DAOS_FAIL_ONCE);
+
+	/* Verify that the DAOS_NVME_FAULTY reaction got triggered. Target should
+	 * be in the DOWN state to trigger rebuild. If the target is still in
+	 * UPINT, attempt the faulty injection again with another target index.
+	 * The NVMe FAULTY device reaction might not have been triggered if the
+	 * target was not assigned to the device.
+	 */
```

Comment:
I don't quite follow this. Shouldn't we just check the 'fail_loc_tgt' or the targets assigned to the device where 'fail_loc_tgt' is assigned to?

Reply:
So now we are passing a specific target upon error injection. The target will be a randomly generated number between 0 and the number of pool targets. If the target idx that is set in the fail loc extra value is not assigned to the current device in the auto_detect_faulty() poller, then the faulty reaction will not be triggered. In that case, the test will increment the target idx and try again with that target. For instance if there are 4 devices on the same node (like in my test cases), then it might take 4 attempts to pass a target that is assigned to the device in the poller. 

Otherwise, I think we get into the case we had before where we would not know what device or what targets got evicted until after-the-fact which makes verification difficult.

Reply:
ah, I see. I think you can just simply remove the DAOS_FAIL_ONCE when setting fail loc. And don't forget to reset the fail loc after it's being triggered.

Reply:
@NiuYawei I tried removing the DAOS_FAIL_ONCE, however the tgt passed does not trigger the faulty device reaction.  I don't think this method works. Not sure how the fail lock is supposed to work w/o DAOS_FAIL_ONCE being set, should it automatically retry? it would need to retry on another tgt idx I would think right?

Reply:
solved by explicitly setting DAOS_FAIL_ALWAYS